engine/crates/runtime/src/cache/cached.rs

```diff
@@ -0,0 +1,895 @@
+    let cached_value = if request_cache_config.cache_control.no_cache {
+        Entry::Miss
+    } else {
+        cache
+            .get(&key)
+            .instrument(tracing::info_span!("cache_get", ray_id = ctx.ray_id()))
+            .await
+            .unwrap_or_else(|e| {
+                tracing::warn!("Error loading {} from cache: {}", key, e);
+                Entry::Miss
+            })
+    };
+
+    match cached_value {
+        Entry::Stale {
+            response,
+            state,
+            is_early_stale,
+        } => {
+            let response = Arc::new(response);
+            let mut revalidated = false;
+
+            // we only want to issue a revalidation if one is not already in progress
+            if state != EntryState::UpdateInProgress {
+                revalidated = true;
+
+                update_stale(
+                    cache,
+                    ctx,
+                    key.clone(),
+                    response.clone(),
+                    value_fut,
+                    config.common_cache_tags.clone(),
+                )
+                .await;
+            }
+
+            tracing::info!(ray_id = ctx.ray_id(), "Cache STALE - {} - {}", revalidated, key);
+
+            // early stales mean early refreshes on our part
+            // they shouldn't be considered as stale from a client perspective
+            if is_early_stale {
+                tracing::info!(ray_id = ctx.ray_id(), "Cache HIT - {}", key);
+                return Ok(CacheResponse::Hit(response));
+            }
+
+            Ok(CacheResponse::Stale {
+                response,
+                updating: revalidated,
+            })
+        }
+        Entry::Hit(gql_response) => {
+            tracing::info!(ray_id = ctx.ray_id(), "Cache HIT - {}", key);
+
+            Ok(CacheResponse::Hit(Arc::new(gql_response)))
+        }
+        Entry::Miss => {
+            tracing::info!(ray_id = ctx.ray_id(), "Cache MISS - {}", key);
+
+            let origin_result = value_fut.await?;
+            let origin_tags = origin_result.cache_tags_with_priority_tags(config.common_cache_tags.clone());
+
+            if origin_result.should_purge_related() {
+                let ray_id = ctx.ray_id().to_string();
+                let cache = Arc::clone(&cache);
+                let purge_cache_tags = origin_tags.clone();
+
+                tracing::info!(ray_id, "Purging global cache by tags: {:?}", purge_cache_tags);
+
+                ctx.wait_until(
+                    async_runtime::make_send_on_wasm(async move {
+                        if let Err(err) = cache
+                            .purge_by_tags(purge_cache_tags)
+                            .instrument(tracing::info_span!("cache_tags_purge", ray_id))
+                            .await
+                        {
+                            tracing::error!(ray_id, "Error global cache purge by tags: {}", err);
```

Comment:
Do we know what order workers-rs runs the wait_until futures in?  Just wondering if this will end up running before or after the wait_until we use for logging...

Reply:
couldn't find any details about `wait_until` future execution. But, looking at our code any `wait_until` future will run before any post request clean up, [here](https://github.com/grafbase/api/blob/main/workers/gateway/src/lib.rs#L97) and [here](https://github.com/grafbase/api/blob/main/workers/executor/src/lib.rs#L84).

Reply:
Ah yes cool - forgot we'd done our own `wait_until` thing on top of the CF one.  Nice one.